### PR TITLE
components: Add types to Icon

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Adds types to the `Icon` component.

It's a little messy, but I think it gets the job done at least (even though we have some ugly type casts). Curious what other folks think... if there's a better way to approach this. It's certainly not ideal.

Only one runtime change but I think it captures the intention of the function better.

The prop types are mostly copied from the DT definition for this component.

## How has this been tested?
Type checks pass.

## Types of changes
New feature.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/components/src/icon/index.tsx [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+type AdditionalProps< T > = T extends ComponentType< infer U >
+	? U
+	: T extends DashiconIcon
+	? SVGProps< SVGSVGElement >
+	: {};
+
+export type Props< P > = BaseProps< P > & AdditionalProps< IconType< P > >;
+
+function Icon< P >( { icon, size, ...additionalProps }: Props< P > ) {
```
Comment (tyxla) [inline packages/components/src/icon/index.tsx]: Removing the default `null` of `icon` seems to be breaking some tests inside the `block-editor` package. Should we revert that change? We'd also need to add `null` as possible values for `icon`, of course.

File: packages/components/src/dashicon/types.ts [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+	| 'welcome-add-page'
+	| 'welcome-comments'
+	| 'welcome-learn-more'
+	| 'welcome-view-site'
+	| 'welcome-widgets-menus'
+	| 'welcome-write-blog'
+	| 'wordpress-alt'
+	| 'wordpress'
+	| 'yes-alt'
+	| 'yes';
```
Comment (tyxla) [inline packages/components/src/dashicon/types.ts]: Where did we get the list from? I can see some icons listed on the public page are missing here, a quick diff shows the following:

* airplane
* amazon
* arrow-up-duplicate
* bank
* beer
* bell
* businessperson
* businesswoman
* calculator
* camera-alt
* car
* code-standards
* coffee
* color-picker
* database
* database-add
* database-export
* database-import
* database-remove
* database-view
* drumstick
* edit-large
* edit-page
* editor-code-duplicate
* food
* fullscreen-alt
* fullscreen-exit-alt
* games
* google
* hourglass
* linkedin
* lock-duplicate
* menu-alt2
* menu-alt3
* money-alt
* open-folder
* pdf
* pets
* pinterest
* plugins-checked
* plus-alt2
* podio
* printer
* privacy
* reddit
* spotify
* superhero
* superhero-alt
* text-page
* twitch
* twitter-alt
* update-alt
* whatsapp
* xing
* youtube

File: packages/components/src/dashicon/types.ts [hunk 3]
```diff
@@ -0,0 +1,290 @@
+export type Icon =
```
Comment (tyxla) [inline packages/components/src/dashicon/types.ts]: Should this be `IconType` or something? `Icon` feels too broad to me.

File: packages/components/src/icon/index.tsx [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+	}
+
+	if ( isValidElement( icon ) && Dashicon === icon.type ) {
+		return cloneElement( icon, {
+			...additionalProps,
+		} );
+	}
+
+	// Icons should be 24x24 by default.
+	const iconSize = size || 24;
```
Comment (tyxla) [inline packages/components/src/icon/index.tsx]: I wonder if would do any harm to specify this as a default arg value instead. Then we could also shorten some of the code by avoiding declaration and referencing an extra variable.

File: packages/components/src/icon/index.tsx (hunk 1 above)
Comment (sarayourfriend) [reply packages/components/src/icon/index.tsx]: Yup! I meant to revert that change anyway 😬

File: packages/components/src/icon/index.tsx (hunk 4 above)
Comment (sarayourfriend) [reply packages/components/src/icon/index.tsx]: Oh yeah, we can do that!

File: packages/components/src/dashicon/types.ts (hunk 2 above)
Comment (sarayourfriend) [reply packages/components/src/dashicon/types.ts]: I took it from the DefinitelyTyped list. I'll go ahead and update it to include the ones you pointed out are missing.

Comment (tyxla) [reply packages/components/src/dashicon/types.ts]: Oh, maybe we should update the DT list as well 😉  The list of available dashicons has been changing with almost every WP version. I think no more time is planned to be invested in them now, though, so the list should be final at this point.

Comment (sarayourfriend) [reply packages/components/src/dashicon/types.ts]: I'll backport this to the DT definitions once it's merged 👍

**Review by ciampo — APPROVED** [review]
I don't have anything to add on top of what @tyxla already commented.

LGTM 🚀

**Pull request merged**